Move from mocha to karma for running unit tests; add code coverage report generation for npm test

clbond opened the pull request:
* Add code coverage report
* Add some missing typings for react-redux-router, redux-thunk, etc
* Add karma dependencies, remove mocha dependencies

clbond commented:
Connected to #112

clbond commented:
@SethDavenport

SethDavenport commented on circle.yml:
```diff
@@ -15,6 +15,7 @@ dependencies:
 
 test:
   post:
+    - cp -R coverage $CIRCLE_ARTIFACTS
```
:+1:

SethDavenport commented on karma.conf.js:
```diff
@@ -0,0 +1,122 @@
+
+  config.set({
+    frameworks: [
+      'jasmine',
```
what is the rationale for karma + jasmine instead of karma + mocha? Just consistency?

Not opposed, just curious

clbond replied:
To be honest I didn't really give it much thought. Our other projects are using Jasmine so I figured I might as well go with that in order to be consistent. But I don't have a strong opinion one way or the other

SethDavenport replied:
I'm cool with it

SethDavenport commented on package.json:
```diff
@@ -76,15 +77,26 @@
     "enzyme": "^2.2.0",
     "es5-shim": "^4.5.6",
     "es6-promise": "^3.1.2",
-    "es6-shim": "^0.35.0",
```
Does babel-polyfill provide everything we need for IE9 support?

e.g. object.assign, array.map/filter/reduce, etc?

The easiest thing would be to make sure the app runs on IE9

clbond commented:
Whoever merges this, please squash the commits together. Some of the earlier commits had nothing to do with this changeset and were later reversed. For clarity's sake it would be better to squash them all together.

SethDavenport commented on package.json:
```diff
@@ -98,13 +110,13 @@
+    "redux-thunk": "^2.1.0",
     "require-hacker": "^2.1.3",
     "rimraf": "^2.4.3",
     "sinon": "^1.17.3",
```
If we've switched to Jasmine we can get rid of sinon - Jasmine has its own spies.

Ditto chai - Jasmine includes its own assert/expect syntax.

SethDavenport commented on src/components/counter/index.tsx:
```diff
@@ -14,19 +15,18 @@ export default function Counter({
 
       <div id="qa-counter-div" className="flex-auto center h1">
-        { counter }
+        {counter.toString()}
```
why is this needed?

clbond replied:
It's not ... good catch. Removed

SethDavenport commented on src/store/configure-store.ts:
```diff
@@ -1,16 +1,22 @@
+import { browserHistory } from 'react-router';
+import { routerMiddleware } from 'react-router-redux';
+
+import thunk from 'redux-thunk';
```
Have the redux-thunk typings been updated? Last time I checked the definity-typed typings were broken for this lib. Specifiy they were incompatible with the official Redux typings.

Granted that was a while ago - hopefully they've updated them.

SethDavenport replied:
Just saw the typing.json changes below - ignor the comment above... redux ecosystem typings continie to improve which is awesome!

SethDavenport commented on typings.json:
```diff
@@ -9,5 +9,10 @@
     "whatwg-fetch": "registry:dt/whatwg-fetch#0.0.0+20160524142046"
+  },
+  "dependencies": {
+    "react-router": "registry:npm/react-router#2.4.0+20160628165748",
```
sweet! official typings

SethDavenport commented:
Overall looks good - if we can get rid of sinon/chai, and verify that it works on IE9 (there will be some CSS issues with the header but that's ok, there's another issue tracking that) then ok to :shipit:

clbond commented:
@SethDavenport Alright, I removed the dependencies on chai and sinon and converted all the assertions to Jasmine syntax. OK to merge once CI finishes?

clbond commented:
@SethDavenport Verified working on IE9. Merging.
